homeassistant/helpers/script.py

```diff
@@ -56,27 +69,23 @@
 ACTION_DEVICE_AUTOMATION = "device"
 ACTION_ACTIVATE_SCENE = "scene"
 
-IF_RUNNING_ERROR = "error"
-IF_RUNNING_IGNORE = "ignore"
-IF_RUNNING_PARALLEL = "parallel"
-IF_RUNNING_RESTART = "restart"
-# First choice is default
-IF_RUNNING_CHOICES = [
-    IF_RUNNING_PARALLEL,
-    IF_RUNNING_ERROR,
-    IF_RUNNING_IGNORE,
-    IF_RUNNING_RESTART,
+SCRIPT_MODE_ERROR = "error"
+SCRIPT_MODE_IGNORE = "ignore"
+SCRIPT_MODE_LEGACY = "legacy"
+SCRIPT_MODE_PARALLEL = "parallel"
+SCRIPT_MODE_QUEUE = "queue"
+SCRIPT_MODE_RESTART = "restart"
+SCRIPT_MODE_CHOICES = [
+    SCRIPT_MODE_ERROR,
+    SCRIPT_MODE_IGNORE,
+    SCRIPT_MODE_LEGACY,
+    SCRIPT_MODE_PARALLEL,
+    SCRIPT_MODE_QUEUE,
+    SCRIPT_MODE_RESTART,
 ]
+DEFAULT_SCRIPT_MODE = SCRIPT_MODE_LEGACY
```

Comment:
We shouldn't specify a default. Let's make sure that we require the caller to specify a mode. That way we avoid surprises. It also allows us to print a warning ahead of changing the default from legacy -> blocking.

Reply:
The `Script` helper class is used in a couple of other places as well (and who knows in how many custom integrations) and I wanted to make sure those continued to work as-is.

I'm handing the warning at the `script` and `automation` integration level. And rather than do it on a per script/automation basis I'm going to only issue one warning (per integration) that provides a list of the entities that haven't specified a mode. I thought one warning would be better than possibly many, many warnings (for people who have a lot of scripts and/or automations.)

Here's how it will look (from my script testing so far):
```
2020-03-03 13:26:42 WARNING (MainThread) [homeassistant.components.script] Script behavior has changed. To continue using previous behavior, which is now deprecated, add 'mode: legacy' to script(s): legacy_cancelable_not, legacy_cancelable_short, legacy_cancelable_long, new_run_new_no_wait_parallel_twice_direct, new_run_new_no_wait_parallel_twice_non_blocking_no_wait, new_run_new_no_wait_parallel_twice_non_blocking_wait, new_run_new_short_parallel_twice_direct, new_run_new_short_parallel_twice_non_blocking_no_wait, new_run_new_short_parallel_twice_non_blocking_wait, legacy_run_legacy_cancelable_not_twice_direct, legacy_run_legacy_cancelable_not_twice_turn_on, legacy_run_legacy_cancelable_short_twice_direct, legacy_run_legacy_cancelable_short_twice_turn_on, legacy_run_legacy_cancelable_long_twice_direct, legacy_run_legacy_cancelable_long_twice_turn_on, legacy_run_new_no_wait_parallel_twice_direct, legacy_run_new_no_wait_parallel_twice_turn_on, 1581351491735, 1581352147150, 1581355326399. See https://www.home-assistant.io/integrations/script/
```

Reply:
The only downside to the warning approach seems to be that we now force people to always specify a mode. We should make sure the frontend does this moving forward.

Reply:
... or live with the warning (until it doesn't work anymore) which is what a lot of people seem to do. 😜

But seriously, we'll only force that until we remove legacy support. (Not sure what you're thinking along those lines.) After that I would expect we'd make the default something else reasonable, possibly depending on the integration. And then there will be no need for a warning because the default will be documented and nothing will be deprecated anymore.